packages/mdc-dialog/foundation.ts

```diff
@@ -205,8 +214,9 @@ export class MDCDialogFoundation extends MDCFoundation<MDCDialogAdapter> {
       return;
     }
 
+    const target = evt.composedPath ? evt.composedPath()[0] : evt.target;
```

Comment:
We still need to address closed shadow roots because the dialog will behave differently if the shadow root is open or closed.

https://jsbin.com/zakopahove/edit?html,output

For these two textarea elements, the target on keypress is `<closed-textarea>` for the first (because composedPath cannot open the closed shadow root), and `<textarea>` for the second (because the shadow root) is open. If we're only looking for a `textarea` selector, we'll match the second, but not the first.

If this was the _same_ element, but the shadow root mode was toggled based on a production vs debug mode flag, that would lead to a bug.

The dialog shouldn't assume that all of its contents will be open shadow roots. Instead, by just viewing `evt.target`, it places the responsibility on the content (aka the ones who _know_ if their selectors need to be updated) to inform the dialog of what the new selectors are.

In the above jsbin example, the content is responsible for informing the dialog that the selector `"closed-textarea, open-textarea"` should suppress the default press.

Reply:
This behaviour is correct. It is closed for a reason, so you can't look into it... as far as we're concerned the correct target we're interested in is the first _visible_ element, i.e. the first entry of the composed path (as it already excludes nodes from closed roots).

im pretty sure this is right, as closed shadow roots are purposely designed so you won't penetrate their boundaries. we definitely should not be aware there is a textarea inside one. it is up to the consumer to instead provide a higher up selector outside the closed root.

Reply:
It's not that the behavior is incorrect on the browser, the semantics of what we're doing is unsafe for the dialog.

`event.composedPath()` has the same penetrative capabilities as `element.shadowRoot.querySelector()`. For open shadow roots it allows external observers to break a shadow root's encapsulation. That's great if you _know_ that all your shadow roots are open and you allow that, but that's something we strictly do not allow with our web components internally.

For MDC's dialog, we'd prefer not to use the `event.composedPath()` API for the same reason that we wouldn't use `element.shadowRoot.querySelector()`. We have use cases where the same element will present with either an open or a closed shadow root depending on the build. If the dialog does not behave the same way for both situations, that can lead to a difficult-to-debug problem (i.e. `<mwc-textarea>` works in debug mode because we're able to penetrate the open shadow root and find its `textarea`, but breaks in production mode because the shadow root is closed).

_We_ could compensate for that with our elements, but then we place that burden on other developers to understand that subtle pattern. It can easily be missed if documentation isn't read carefully. 

The safer option is to avoid shadow root penetrating APIs and require opaque selectors for all web components, regardless of whether or not their shadow root is open or closed.

Reply:
i still disagree. look at this from a functional point of view, we want to detect _visible_ targets key events have originated from, where "visible" is in terms of DOM tree exposure. we want to then "do nothing" when one of those origins is a `textarea`. The way to do that is exactly what we have here.

Everything you've described is because you're trying to account for implementation detail in how you've chosen to structure your tests, completely unrelated to the change right here.

Closed shadow roots are purposely impossible to traverse from outside (AFAIK), there is no way to know if one exists and no way to penetrate the boundary of one. So it really isn't any concern of ours to care for the purposes of the code here.

> The safer option is to avoid shadow root penetrating APIs and require opaque selectors for all web components, regardless of whether or not their shadow root is open or closed.

This would re-introduce the problem we are currently trying to fix. We **want** to detect elements in shadow roots here because it is the correct thing to do and the reason this fix and its related issue exists at all. 

Again, this sounds to me like you're trying to apply restrictions to what would otherwise be a valid fix because of how your tests are currently executed/structured around it. If that's the case, the tests are wrong, not the fix. This is specifically a feature to penetrate shadow roots, on purpose. Keep in mind the usage of closed shadow roots is incredibly low, and anyone who chooses to use one im sure is very capable of accounting for the fact that _the platform_ will hide their root's children from _everything_.

Reply:
Let's take an example. I have a dialog with a textarea in it, 4 shadow roots deep.

With this, i can safely know that the primary action will not be triggered because i am in a textarea. If i have other elements i want to be able to press "enter" in, i can change the selector to also match those.

Let's now assume this doesn't exist. Let's assume you came up with some other solution where we are more explicit (maybe we pass a handler in, and deal with textarea detection ourselves as the consumer). You still will never be able to detect or enter a closed shadow root (as defined by the browser), so neither can a consumer. So how exactly do you expect them to ever be capable of preventing a textarea in their closed root from triggering the primary action?

It simply should not be possible and is not possible as far as i know, because thats what a closed shadow root is meant to behave like. Given that, the only possible solution for someone with a closed root seems to be to use a selector/element ABOVE the closed root. Which is exactly where we're at here.

Reply:
Thanks for the explanation!

I discussed this with the team, and while they agree that we generally want to avoid `composedPath()`, the argument for nested shadow root content is a good one considering that using web components as a "layout" element is not uncommon.

```
<mwc-dialog>
  <horizontal-layout>
    #shadow-root
      <mwc-textarea></mwc-textarea>
      <mwc-checkbox></mwc-checkbox>
  </horizontal-layout>
</mwc-dialog>
```

In the above scenario, only pressing enter while focused in the textarea should prevent the dialog from closing, which would not be possible unless the dialog penetrates the shadow root (the user would have to instruct the dialog that any enter event from `<horizontal-layout>` would prevent the dialog from closing).

To wrap up this discussion, let's add a comment so we can capture why we're using `composedPath()` here:

```
// Prefer composedPath() (if available) to retrieve the original target from any open shadow roots.
// This avoids problems when using web components in the dialog's content
// <div class="mdc-dialog__content">
//   <horizontal-layout>
//     #shadow-root
//       <textarea></textarea>
//   </horizontal-layout>
// </div>
const target = evt.composedPath ? evt.composedPath()[0] : evt.target;
```